c/pthread-C-DAC/pthread-producer-consumer.c

```diff
@@ -0,0 +1,227 @@
+ 
+  Example     : pthread-producer-consumer.c
+  
+  Objective   : To illustrate producer-consumer problem using pthreads.
+ 
+  Input       : Nothing.
+ 
+  Output      : Sequence of Produced and Consumed items.
+                                                                                                                         
+  Created     : MAY-2013 
+  E-mail      : hpcfte@cdac.in          
+                                
+****************************************************************************/
+ 
+#include <pthread.h>
+#include <stdio.h>
+#include <unistd.h>
+#include <stdlib.h>
+ 
+/*defining QUEUE sise */
+#define QUEUESIZE 10
+#define LOOP 20
+ 
+/*global variable declaration */
+int ret_count;
+/*Thread callback function prototype*/
+void *producer (void *args);
+void *consumer (void *args);
+ 
+typedef struct {
+         
+            int buf[QUEUESIZE];
+            long head, tail;
+            int full, empty;
+            pthread_mutex_t *mut;
+            pthread_cond_t *notFull, *notEmpty;
+           } queue;
+ 
+ 
+ 
+queue *queueInit (void);
+void queueDelete (queue *q);
+void queueAdd (queue *q, int in);
+void queueDel (queue *q, int *out);
+ 
+/*Main function */
+int main ()
+{
+    queue *fifo;
+    pthread_t pro, con;
+    /*QUEUE initialization */
+    fifo = queueInit ();
+    if (fifo == NULL)
+        {
+        fprintf (stderr, "main: Queue Init failed.\n");
+        exit (1);
+    }
+    /*producer thread creation */
+    ret_count=pthread_create (&pro, NULL, producer, fifo);
+        if(ret_count)
+        {
+        printf("\n ERROR : Return code from pthread_create() is %d ",ret_count);
+        exit(-1);
+        }
+    /*consumer thread creation */
+    ret_count=pthread_create (&con, NULL, consumer, fifo);
+    if(ret_count)
+        {
+        printf("\n ERROR : Return code from pthread_create() is %d ",ret_count);
+        exit(-1);
+        }
+    /*joining producer thread to main thread */
+    ret_count=pthread_join (pro, NULL);
+        if(ret_count)
+        {
+        printf("\n ERROR : Return code from pthread_join() is %d ",ret_count);
+        exit(-1);
+        }
+ 
+ 
+    /*joining consumer thread to main thread */
+    ret_count=pthread_join (con, NULL);
+    if(ret_count)
+        {
+        printf("\n ERROR : Return code from pthread_join() is %d ",ret_count);
+        exit(-1);
+        }
+    queueDelete (fifo);
+    return 0;
+}
+ 
+/*producer thread callback function */
+void *producer (void *q)
+{
+    queue *fifo;
+    int i;
+    fifo = (queue *)q;
+    for (i = 0; i < LOOP; i++)
+    {
+        pthread_mutex_lock (fifo->mut);
+        while (fifo->full)
+        {
+            printf ("producer: queue FULL.\n");
+            pthread_cond_wait (fifo->notFull, fifo->mut);
+        }
+        queueAdd (fifo, i);
+        pthread_mutex_unlock (fifo->mut);
+        pthread_cond_signal (fifo->notEmpty);
+        printf ("producer: add %d \n",i);
+        usleep (100000);
```

Comment:
Does SVCOMP support sleeping? Or any other assumptions about timing?
If the function "usleep(...)" is irrelevant for the property, it should either be removed or a simple definition should be given.